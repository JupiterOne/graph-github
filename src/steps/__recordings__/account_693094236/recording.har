{
  "log": {
    "_recordingName": "account",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "6.0.5"
    },
    "entries": [
      {
        "_id": "f13f3f3dcf87065d27af1cac191fa6e2",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/vnd.github.machine-man-preview+json"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "jupiter-integration-github octokit-rest.js/19.0.5 octokit-core.js/4.1.0 Node.js/18.16.0 (darwin; x64)"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "0"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.github.com"
            }
          ],
          "headersSize": 812,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "queryString": [],
          "url": "https://api.github.com/app/installations/32841752/access_tokens"
        },
        "response": {
          "bodySize": 493,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 493,
            "text": "{\"token\":\"[REDACTED]\",\"expires_at\":\"2050-12-31T18:09:20Z\",\"permissions\":{\"members\":\"read\",\"organization_administration\":\"read\",\"organization_secrets\":\"read\",\"actions\":\"read\",\"administration\":\"write\",\"checks\":\"read\",\"discussions\":\"read\",\"environments\":\"read\",\"issues\":\"read\",\"metadata\":\"read\",\"pages\":\"read\",\"pull_requests\":\"read\",\"repository_advisories\":\"read\",\"secrets\":\"read\",\"security_events\":\"read\",\"vulnerability_alerts\":\"read\"},\"repository_selection\":\"all\"}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "GitHub.com"
            },
            {
              "name": "date",
              "value": "Mon, 24 Jul 2023 16:54:47 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "content-length",
              "value": "493"
            },
            {
              "name": "cache-control",
              "value": "public, max-age=60, s-maxage=60"
            },
            {
              "name": "vary",
              "value": "Accept, Accept-Encoding, Accept, X-Requested-With"
            },
            {
              "name": "etag",
              "value": "\"ad52821a27fbc9a0c4e985e46949c90c4bd54cdab17854073c704a1598ff7c60\""
            },
            {
              "name": "x-github-media-type",
              "value": "github.v3; param=machine-man-preview; format=json"
            },
            {
              "name": "x-github-api-version-selected",
              "value": "2022-11-28"
            },
            {
              "name": "access-control-expose-headers",
              "value": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubdomains; preload"
            },
            {
              "name": "x-frame-options",
              "value": "deny"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "referrer-policy",
              "value": "origin-when-cross-origin, strict-origin-when-cross-origin"
            },
            {
              "name": "content-security-policy",
              "value": "default-src 'none'"
            },
            {
              "name": "x-github-request-id",
              "value": "2055:238A:23C7FD:4806A9:64BEACD7"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 1114,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 201,
          "statusText": "Created"
        },
        "startedDateTime": "2023-07-24T16:54:46.554Z",
        "time": 509,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 509
        }
      },
      {
        "_id": "9ca046c0c3e008801800630bedd7bdf3",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/vnd.github.v3+json"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "jupiter-integration-github octokit-rest.js/19.0.5 octokit-core.js/4.1.0 Node.js/18.16.0 (darwin; x64)"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.github.com"
            }
          ],
          "headersSize": 761,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://api.github.com/app/installations/32841752"
        },
        "response": {
          "bodySize": 959,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 959,
            "text": "{\"id\":32841752,\"account\":{\"login\":\"JupiterOne-Sandbox\",\"id\":121981149,\"node_id\":\"O_kgDOB0VI3Q\",\"avatar_url\":\"https://avatars.githubusercontent.com/u/121981149?v=4\",\"gravatar_id\":\"\",\"url\":\"https://api.github.com/users/JupiterOne-Sandbox\",\"html_url\":\"https://github.com/JupiterOne-Sandbox\",\"followers_url\":\"https://api.github.com/users/JupiterOne-Sandbox/followers\",\"following_url\":\"https://api.github.com/users/JupiterOne-Sandbox/following{/other_user}\",\"gists_url\":\"https://api.github.com/users/JupiterOne-Sandbox/gists{/gist_id}\",\"starred_url\":\"https://api.github.com/users/JupiterOne-Sandbox/starred{/owner}{/repo}\",\"subscriptions_url\":\"https://api.github.com/users/JupiterOne-Sandbox/subscriptions\",\"organizations_url\":\"https://api.github.com/users/JupiterOne-Sandbox/orgs\",\"repos_url\":\"https://api.github.com/users/JupiterOne-Sandbox/repos\",\"events_url\":\"https://api.github.com/users/JupiterOne-Sandbox/events{/privacy}\",\"received_events_url\":\"https://api.github.com/users/JupiterOne-Sandbox/received_events\",\"type\":\"Organization\",\"site_admin\":false},\"repository_selection\":\"all\",\"access_tokens_url\":\"https://api.github.com/app/installations/32841752/access_tokens\",\"repositories_url\":\"https://api.github.com/installation/repositories\",\"html_url\":\"https://github.com/organizations/JupiterOne-Sandbox/settings/installations/32841752\",\"app_id\":278393,\"app_slug\":\"j1-ingest-cg\",\"target_id\":121981149,\"target_type\":\"Organization\",\"permissions\":{\"pages\":\"read\",\"checks\":\"read\",\"issues\":\"read\",\"actions\":\"read\",\"members\":\"read\",\"secrets\":\"read\",\"metadata\":\"read\",\"discussions\":\"read\",\"environments\":\"read\",\"pull_requests\":\"read\",\"administration\":\"write\",\"security_events\":\"read\",\"organization_secrets\":\"read\",\"vulnerability_alerts\":\"read\",\"repository_advisories\":\"read\",\"organization_administration\":\"read\"},\"events\":[],\"created_at\":\"2023-01-05T16:02:41.000Z\",\"updated_at\":\"2023-05-23T15:42:13.000Z\",\"single_file_name\":null,\"has_multiple_single_files\":false,\"single_file_paths\":[],\"suspended_by\":null,\"suspended_at\":null}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "GitHub.com"
            },
            {
              "name": "date",
              "value": "Mon, 24 Jul 2023 16:54:47 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "cache-control",
              "value": "public, max-age=60, s-maxage=60"
            },
            {
              "name": "vary",
              "value": "Accept, Accept-Encoding, Accept, X-Requested-With"
            },
            {
              "name": "etag",
              "value": "W/\"a88ec880b48eb9520cef4f46a213cd73637d94f4c441e4d5954b938ccd1f2376\""
            },
            {
              "name": "x-github-media-type",
              "value": "github.v3; format=json"
            },
            {
              "name": "x-github-api-version-selected",
              "value": "2022-11-28"
            },
            {
              "name": "access-control-expose-headers",
              "value": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubdomains; preload"
            },
            {
              "name": "x-frame-options",
              "value": "deny"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "referrer-policy",
              "value": "origin-when-cross-origin, strict-origin-when-cross-origin"
            },
            {
              "name": "content-security-policy",
              "value": "default-src 'none'"
            },
            {
              "name": "x-github-request-id",
              "value": "205A:45BC:222C7C:44DFDA:64BEACD7"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 1120,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2023-07-24T16:54:47.095Z",
        "time": 322,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 322
        }
      },
      {
        "_id": "22e4c882794d943e8a6f0961702d5000",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 452,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/vnd.github.v3+json"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "jupiterone-graph-github"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "452"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.github.com"
            }
          ],
          "headersSize": 328,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "postData": {
            "mimeType": "application/json; charset=utf-8",
            "params": [],
            "text": "{\"query\":\"\\n    query ($login: String!) {\\n      organization(login: $login) {\\n        id\\n        login\\n        name\\n        createdAt\\n        updatedAt\\n        description\\n        email\\n        databaseId\\n        isVerified\\n        location\\n        websiteUrl\\n        url\\n      }\\n      ...on Query {\\n    rateLimit {\\n      limit\\n      cost\\n      remaining\\n      resetAt\\n    }\\n  }\\n    }\",\"variables\":{\"login\":\"JupiterOne-Sandbox\"}}"
          },
          "queryString": [],
          "url": "https://api.github.com/graphql"
        },
        "response": {
          "bodySize": 427,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 427,
            "text": "{\"data\":{\"organization\":{\"id\":\"O_kgDOB0VI3Q\",\"login\":\"JupiterOne-Sandbox\",\"name\":\"JupiterOne Sandbox\",\"createdAt\":\"2023-01-04T21:30:17Z\",\"updatedAt\":\"2023-03-17T20:49:58Z\",\"description\":\"\",\"email\":\"hello@jupiterone.com\",\"databaseId\":121981149,\"isVerified\":false,\"location\":\"United States of America\",\"websiteUrl\":\"https://jupiterone.com\",\"url\":\"https://github.com/JupiterOne-Sandbox\"},\"rateLimit\":{\"limit\":5000,\"cost\":1,\"remaining\":5000,\"resetAt\":\"2023-07-24T17:19:11Z\"}}}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "GitHub.com"
            },
            {
              "name": "date",
              "value": "Mon, 24 Jul 2023 16:54:48 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "x-github-media-type",
              "value": "github.v3; format=json"
            },
            {
              "name": "x-ratelimit-limit",
              "value": "5000"
            },
            {
              "name": "x-ratelimit-remaining",
              "value": "5000"
            },
            {
              "name": "x-ratelimit-reset",
              "value": "1690219151"
            },
            {
              "name": "x-ratelimit-used",
              "value": "805"
            },
            {
              "name": "x-ratelimit-resource",
              "value": "graphql"
            },
            {
              "name": "access-control-expose-headers",
              "value": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubdomains; preload"
            },
            {
              "name": "x-frame-options",
              "value": "deny"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "referrer-policy",
              "value": "origin-when-cross-origin, strict-origin-when-cross-origin"
            },
            {
              "name": "content-security-policy",
              "value": "default-src 'none'"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, Accept, X-Requested-With"
            },
            {
              "name": "x-github-request-id",
              "value": "205D:3152:23CFB9:4818E4:64BEACD7"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 1084,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2023-07-24T16:54:47.430Z",
        "time": 391,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 391
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
