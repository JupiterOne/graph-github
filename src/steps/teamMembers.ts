import {
  IntegrationLogger,
  IntegrationMissingKeyError,
  IntegrationStep,
  IntegrationStepExecutionContext,
  JobState,
  RelationshipClass,
  createDirectRelationship,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../config';
import { toOrganizationMemberEntityFromTeamMember } from '../sync/converters';
import {
  OrgTeamMemberQueryResponse,
  TeamMemberRole,
  getOrCreateGraphqlClient,
} from '../client/GraphQLClient';
import {
  GithubEntities,
  Steps,
  Relationships,
  MEMBERS_TOTAL_BY_TEAM,
  TEAM_DATA_MAP,
} from '../constants';
import { TeamData } from '../types';
import { withBatching } from '../client/GraphQLClient/batchUtils';

export async function fetchTeamMembers({
  instance,
  logger,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const config = instance.config;
  const graphqlClient = getOrCreateGraphqlClient(config, logger);

  const teamDataMap =
    await jobState.getData<Map<string, TeamData>>(TEAM_DATA_MAP);
  if (!teamDataMap) {
    throw new IntegrationMissingKeyError(
      `Expected teams.ts to have set ${TEAM_DATA_MAP} in jobState.`,
    );
  }

  const membersTotalByRepo = await jobState.getData<Map<string, number>>(
    MEMBERS_TOTAL_BY_TEAM,
  );
  if (!membersTotalByRepo) {
    return;
  }

  const iteratee = buildIteratee({ jobState, logger, config });

  await withBatching({
    totalConnectionsById: membersTotalByRepo,
    threshold: 100,
    batchCb: async (teamKeys) => {
      await graphqlClient.iterateTeamMembers(teamKeys, iteratee);
    },
    singleCb: async (teamKey) => {
      const teamData = teamDataMap.get(teamKey);
      if (!teamData) {
        return;
      }
      await graphqlClient.iterateTeamMembers(teamData.name, iteratee);
    },
    logger,
  });

  await jobState.deleteData(MEMBERS_TOTAL_BY_TEAM);
}

const buildIteratee = ({
  jobState,
  logger,
  config,
}: {
  jobState: JobState;
  logger: IntegrationLogger;
  config: IntegrationConfig;
}) => {
  return async (user: OrgTeamMemberQueryResponse) => {
    if (!jobState.hasKey(user.id)) {
      //somehow this team has a user we didn't know about
      //shouldn't happen, except through weird timing, but we'll make an entry
      await jobState.addEntity(
        toOrganizationMemberEntityFromTeamMember(user, config.githubApiBaseUrl),
      );
    }

    const teamMemberRelationship = createDirectRelationship({
      _class: RelationshipClass.HAS,
      fromKey: user.teamId,
      fromType: GithubEntities.GITHUB_TEAM._type,
      toKey: user.id,
      toType: GithubEntities.GITHUB_MEMBER._type,
    });

    if (jobState.hasKey(teamMemberRelationship._key)) {
      logger.warn(
        {
          teamId: user.teamId,
          teamName: user.teamName,
          teamMemberId: user.id,
          relationshipKey: teamMemberRelationship._key,
        },
        'Member-team relationship was already ingested: Skipping.',
      );
    } else {
      await jobState.addRelationship(teamMemberRelationship);
    }

    if (user.role === TeamMemberRole.Maintainer) {
      const maintainerTeamRelationship = createDirectRelationship({
        _class: RelationshipClass.MANAGES,
        fromKey: user.id,
        fromType: GithubEntities.GITHUB_MEMBER._type,
        toKey: user.teamId,
        toType: GithubEntities.GITHUB_TEAM._type,
      });

      if (jobState.hasKey(maintainerTeamRelationship._key)) {
        logger.warn(
          {
            teamId: user.teamId,
            teamName: user.teamName,
            teamMemberId: user.id,
            relationshipKey: maintainerTeamRelationship._key,
          },
          'Maintainer-team relationship was already ingested: Skipping.',
        );
      } else {
        await jobState.addRelationship(maintainerTeamRelationship);
      }
    }
  };
};

export const teamMemberSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: Steps.FETCH_TEAM_MEMBERS,
    name: 'Fetch Team Members',
    entities: [],
    relationships: [
      Relationships.TEAM_HAS_USER,
      Relationships.USER_MANAGES_TEAM,
    ],
    dependsOn: [Steps.FETCH_TEAMS, Steps.FETCH_USERS],
    executionHandler: fetchTeamMembers,
  },
];
