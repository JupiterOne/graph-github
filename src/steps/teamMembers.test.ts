import { Recording } from '@jupiterone/integration-sdk-testing';
import { sanitizeConfig } from '../config';
import { teamMemberSteps } from './teamMembers';
import { integrationConfig } from '../../test/config';
import { setupGithubRecording } from '../../test/recording';
import { Relationships } from '../constants';
import { invocationConfig } from '..';
import { executeStepWithDependencies } from '../../test/executeStepWithDependencies';

jest.setTimeout(20000);

let recording: Recording;
afterEach(async () => {
  await recording.stop();
});

test('fetchTeamMembers exec handler', async () => {
  recording = setupGithubRecording({
    directory: __dirname,
    name: 'teamMembers',
  });
  sanitizeConfig(integrationConfig);

  const { collectedEntities, collectedRelationships, encounteredTypes } =
    await executeStepWithDependencies({
      stepId: teamMemberSteps[0].id,
      invocationConfig: invocationConfig as any,
      instanceConfig: integrationConfig,
    });

  expect({
    numCollectedEntities: collectedEntities.length,
    numCollectedRelationships: collectedRelationships.length,
    collectedEntities: collectedEntities,
    collectedRelationships: collectedRelationships,
    encounteredTypes: encounteredTypes,
  }).toMatchSnapshot();

  // this step only makes relationships
  const teamHasMemberRels = collectedRelationships.filter(
    (e) => e._type === Relationships.TEAM_HAS_USER._type,
  );
  expect(teamHasMemberRels.length).toBeGreaterThan(0);

  const memberManagesTeamRels = collectedRelationships.filter(
    (e) => e._type === Relationships.USER_MANAGES_TEAM._type,
  );
  expect(memberManagesTeamRels.length).toBeGreaterThan(0);
});
