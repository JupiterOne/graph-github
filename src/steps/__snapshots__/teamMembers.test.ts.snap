// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchTeamMembers exec handler 1`] = `
Object {
  "collectedEntities": Array [],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "T_kwDOB0VI3c4Ab0na",
      "_key": "T_kwDOB0VI3c4Ab0na|has|MDQ6VXNlcjM3NzE5MjQ=",
      "_toEntityKey": "MDQ6VXNlcjM3NzE5MjQ=",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjM3NzE5MjQ=",
      "_key": "MDQ6VXNlcjM3NzE5MjQ=|manages|T_kwDOB0VI3c4Ab0na",
      "_toEntityKey": "T_kwDOB0VI3c4Ab0na",
      "_type": "github_user_manages_team",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "T_kwDOB0VI3c4Ab0na",
      "_key": "T_kwDOB0VI3c4Ab0na|has|MDQ6VXNlcjU5MjMwMDI=",
      "_toEntityKey": "MDQ6VXNlcjU5MjMwMDI=",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjU5MjMwMDI=",
      "_key": "MDQ6VXNlcjU5MjMwMDI=|manages|T_kwDOB0VI3c4Ab0na",
      "_toEntityKey": "T_kwDOB0VI3c4Ab0na",
      "_type": "github_user_manages_team",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "T_kwDOB0VI3c4Ab0na",
      "_key": "T_kwDOB0VI3c4Ab0na|has|MDQ6VXNlcjcxOTA2OTY=",
      "_toEntityKey": "MDQ6VXNlcjcxOTA2OTY=",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjcxOTA2OTY=",
      "_key": "MDQ6VXNlcjcxOTA2OTY=|manages|T_kwDOB0VI3c4Ab0na",
      "_toEntityKey": "T_kwDOB0VI3c4Ab0na",
      "_type": "github_user_manages_team",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "T_kwDOB0VI3c4Ab0na",
      "_key": "T_kwDOB0VI3c4Ab0na|has|MDQ6VXNlcjU1NzcxODIz",
      "_toEntityKey": "MDQ6VXNlcjU1NzcxODIz",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjU1NzcxODIz",
      "_key": "MDQ6VXNlcjU1NzcxODIz|manages|T_kwDOB0VI3c4Ab0na",
      "_toEntityKey": "T_kwDOB0VI3c4Ab0na",
      "_type": "github_user_manages_team",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "T_kwDOB0VI3c4Ab0na",
      "_key": "T_kwDOB0VI3c4Ab0na|has|U_kgDOBaVAgQ",
      "_toEntityKey": "U_kgDOBaVAgQ",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "T_kwDOB0VI3c4AeONM",
      "_key": "T_kwDOB0VI3c4AeONM|has|MDQ6VXNlcjE2NTYyMDg1",
      "_toEntityKey": "MDQ6VXNlcjE2NTYyMDg1",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjE2NTYyMDg1",
      "_key": "MDQ6VXNlcjE2NTYyMDg1|manages|T_kwDOB0VI3c4AeONM",
      "_toEntityKey": "T_kwDOB0VI3c4AeONM",
      "_type": "github_user_manages_team",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "T_kwDOB0VI3c4AeONM",
      "_key": "T_kwDOB0VI3c4AeONM|has|MDQ6VXNlcjMxMjE4NTA3",
      "_toEntityKey": "MDQ6VXNlcjMxMjE4NTA3",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjMxMjE4NTA3",
      "_key": "MDQ6VXNlcjMxMjE4NTA3|manages|T_kwDOB0VI3c4AeONM",
      "_toEntityKey": "T_kwDOB0VI3c4AeONM",
      "_type": "github_user_manages_team",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "T_kwDOB0VI3c4AeONM",
      "_key": "T_kwDOB0VI3c4AeONM|has|U_kgDOBaVAgQ",
      "_toEntityKey": "U_kgDOBaVAgQ",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "github_account",
    "github_team",
    "github_user",
    "github_account_has_team",
    "github_account_has_user",
    "github_user_manages_account",
    "github_team_has_user",
    "github_user_manages_team",
  ],
  "numCollectedEntities": 0,
  "numCollectedRelationships": 14,
}
`;
